### ISSUE TITLE
多标签训练异常

### ISSUE THREAD

**Issue by lergliu** [open-mmlab/mmclassification#351]
基于vgg16_b16x8_voc.py，训练多标签分类模型，其中一个标签的类别数量超过2，训练不报错，但是loss一直为nan。把这个标签去掉，，其余标签的类别数量为2，loss就正常了。现在的多标签版本是不是不支持类别数量超过2的呢？

**Comment by mzr1996** [comment]
可以提供一下训练的配置文件和训练日志吗？

**Comment by lergliu** [comment]
------------------------------------------------------------
sys.platform: linux
Python: 3.7.7 (default, May  7 2020, 21:25:33) [GCC 7.3.0]
CUDA available: True
GPU 0,1,2,3: Tesla V100-PCIE-32GB
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 10.0, V10.0.130
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.6.0
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.1 Product Build 20200208 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.5.0 (Git Hash e2ac1fac44c5078ca927cb9b90e1b3066a0b2ed0)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,

TorchVision: 0.7.0
OpenCV: 4.4.0
MMCV: 1.2.6
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 10.1
MMClassification: 0.7.0+
------------------------------------------------------------

2021-07-08 07:09:42,676 - mmcls - INFO - Distributed training: False
2021-07-08 07:09:43,007 - mmcls - INFO - Config:
dataset_type = 'Face'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='RandomResizedCrop', size=224),
    dict(type='RandomFlip', flip_prob=0.5, direction='horizontal'),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='ToTensor', keys=['gt_label']),
    dict(type='Collect', keys=['img', 'gt_label'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='Resize', size=(256, -1)),
    dict(type='CenterCrop', crop_size=224),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='Collect', keys=['img'])
]
data = dict(
    samples_per_gpu=16,
    workers_per_gpu=2,
    train=dict(
        type='Face',
        data_prefix='data/face/img_align_celeba/',
        ann_file='data/face/val_normed.csv',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='RandomResizedCrop', size=224),
            dict(type='RandomFlip', flip_prob=0.5, direction='horizontal'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='ToTensor', keys=['gt_label']),
            dict(type='Collect', keys=['img', 'gt_label'])
        ]),
    val=dict(
        type='Face',
        data_prefix='data/face/img_align_celeba/',
        ann_file='data/face/val_normed.csv',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='Resize', size=(256, -1)),
            dict(type='CenterCrop', crop_size=224),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ]),
    test=dict(
        type='Face',
        data_prefix='data/face/img_align_celeba/',
        ann_file='data/face/val_normed.csv',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='Resize', size=(256, -1)),
            dict(type='CenterCrop', crop_size=224),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ]))
evaluation = dict(
    interval=1, metric=['mAP', 'CP', 'OP', 'CR', 'OR', 'CF1', 'OF1'])
checkpoint_config = dict(interval=1)
log_config = dict(interval=100, hooks=[dict(type='TextLoggerHook')])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = 'checkpoints/vgg16_batch256_imagenet_20210208-db26f1a5.pth'
resume_from = None
workflow = [('train', 1)]
model = dict(
    type='ImageClassifier',
    backbone=dict(type='VGG', depth=16, num_classes=2),
    neck=None,
    head=dict(
        type='MultiLabelClsHead',
        loss=dict(type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0)))
optimizer = dict(
    type='SGD',
    lr=0.001,
    momentum=0.9,
    weight_decay=0,
    paramwise_cfg=dict(
        custom_keys=dict({'.backbone.classifier': dict(lr_mult=10)})))
optimizer_config = dict(grad_clip=None)
lr_config = dict(policy='step', step=20, gamma=0.1)
runner = dict(type='EpochBasedRunner', max_epochs=40)
work_dir = './work_dirs/vgg16_b16x8_face'
gpu_ids = range(0, 1)

2021-07-08 07:09:55,165 - mmcls - INFO - load checkpoint from checkpoints/vgg16_batch256_imagenet_20210208-db26f1a5.pth
2021-07-08 07:09:55,948 - mmcls - WARNING - The model and loaded state dict do not match exactly

size mismatch for backbone.classifier.6.weight: copying a param with shape torch.Size([1000, 4096]) from checkpoint, the shape in current model is torch.Size([2, 4096]).
size mismatch for backbone.classifier.6.bias: copying a param with shape torch.Size([1000]) from checkpoint, the shape in current model is torch.Size([2]).
2021-07-08 07:09:56,067 - mmcls - INFO - Start running, host: root@6bcd8739ac6c, work_dir: /home/ailab/mmclassification/work_dirs/vgg16_b16x8_face
2021-07-08 07:09:56,067 - mmcls - INFO - workflow: [('train', 1)], max: 40 epochs
2021-07-08 07:10:07,924 - mmcls - INFO - Epoch [1][100/1250]    lr: 1.000e-03, eta: 1:38:05, time: 0.118, data_time: 0.028, memory: 3053, loss: nan
2021-07-08 07:10:16,806 - mmcls - INFO - Epoch [1][200/1250]    lr: 1.000e-03, eta: 1:25:48, time: 0.089, data_time: 0.006, memory: 3053, loss: nan
2021-07-08 07:10:26,091 - mmcls - INFO - Epoch [1][300/1250]    lr: 1.000e-03, eta: 1:22:43, time: 0.093, data_time: 0.006, memory: 3053, loss: nan

**Comment by mzr1996** [comment]
好的，我们会在近期进行检查

**Comment by lergliu** [comment]
> 好的，我们会在近期进行检查

谢谢

**Comment by mzr1996** [comment]
您好，可以提供一下Face数据集类的实现吗？

**Comment by lergliu** [comment]
> 您好，可以提供一下Face数据集类的实现吗？

class Face(MultiLabelDataset):

    CLASSES = {'hair_color':['black', 'gray', 'yellow', 'brown', '-1'],
               'gender': ['female', 'male'],
               'age': ['young', 'old']}

    def load_annotations(self):

        data_infos = []
        with open(self.ann_file) as f:
            reader = csv.DictReader(f)
            for row in reader:
                filename = row['image_path'].split('/')[-1]
                labels = []
                for (key, value) in CLASSES.items():
                    attr = row[key]
                    label = value.index(attr)
                    labels.append(label)
                
                info = dict(
                    img_prefix=self.data_prefix,
                    img_info=dict(filename=filename),
                    gt_label=np.array(labels).astype(np.int8))
                data_infos.append(info)
        return data_infos

数据集是csv格式的，格式如下图
![11111](https://user-images.githubusercontent.com/62641051/125014070-90482080-e09f-11eb-9745-f2ca46068eda.png)

**Comment by mzr1996** [comment]
这里有三处错误，请尝试对以下问题进行修正，再测试是否出现 nan 问题
```python
            labels = []
            for (key, value) in CLASSES.items():
                attr = row[key]
                label = value.index(attr)
                labels.append(label)
```
这里，你对不同 field 分别进行了标签化操作，但没有添加 offset，如果一个人的类别是 black, famale, young, 那么产生的 `labels`将会是`[0, 0, 0]`，这是不合理的。
建议为后两个 field 增加 offset，使其`labels`变为`[0, 5, 7]`

---
```python
                gt_label=np.array(labels).astype(np.int8))
```
这里，对于多分类，`gt_label`应该以 one-hot 的形式保存，如上例中的 `[0, 5, 7]` 的 `gt_label` 应当保存为 `[1, 0, 0, 0, 0, 1, 0, 1, 0]`，具体实现可以参考 `VOC` 数据集类。

---
```python
backbone=dict(type='VGG', depth=16, num_classes=2),
```
最后，这里的 `num_classes` 表示的不是有多少 field，而是具体有多少类别需要预测，这里应该是 9，包括 hair_color 的 5 类，gender 的 2 类，age 的 2 类。

**Comment by lergliu** [comment]
> 这里有三处错误，请尝试对以下问题进行修正，再测试是否出现 nan 问题
> 
> ```python
>             labels = []
>             for (key, value) in CLASSES.items():
>                 attr = row[key]
>                 label = value.index(attr)
>                 labels.append(label)
> ```
> 
> 这里，你对不同 field 分别进行了标签化操作，但没有添加 offset，如果一个人的类别是 black, famale, young, 那么产生的 `labels`将会是`[0, 0, 0]`，这是不合理的。
> 建议为后两个 field 增加 offset，使其`labels`变为`[0, 5, 7]`
> 
> ```python
>                 gt_label=np.array(labels).astype(np.int8))
> ```
> 
> 这里，对于多标签分类，`gt_label`应该以 one-hot 的形式保存，如上例中的 `[0, 5, 7]` 的 `gt_label` 应当保存为 `[1, 0, 0, 0, 0, 1, 0, 1, 0]`，具体实现可以参考 `VOC` 数据集类。
> 
> ```python
> backbone=dict(type='VGG', depth=16, num_classes=2),
> ```
> 
> 最后，这里的 `num_classes` 表示的不是有多少 field，而是具体有多少类别需要预测，这里应该是 9，包括 hair_color 的 5 类，gender 的 2 类，age 的 2 类。

已修改，目前已经能正常训练了。谢谢！

**Comment by hjw-0909** [comment]
> > 这里有三处错误，请尝试对以下问题进行修正，再测试是否出现 nan 问题
> > ```python
> >             labels = []
> >             for (key, value) in CLASSES.items():
> >                 attr = row[key]
> >                 label = value.index(attr)
> >                 labels.append(label)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 这里，你对不同 field 分别进行了标签化操作，但没有添加 offset，如果一个人的类别是 black, famale, young, 那么产生的 `labels`将会是`[0, 0, 0]`，这是不合理的。
> > 建议为后两个 field 增加 offset，使其`labels`变为`[0, 5, 7]`
> > ```python
> >                 gt_label=np.array(labels).astype(np.int8))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 这里，对于多标签分类，`gt_label`应该以 one-hot 的形式保存，如上例中的 `[0, 5, 7]` 的 `gt_label` 应当保存为 `[1, 0, 0, 0, 0, 1, 0, 1, 0]`，具体实现可以参考 `VOC` 数据集类。
> > ```python
> > backbone=dict(type='VGG', depth=16, num_classes=2),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 最后，这里的 `num_classes` 表示的不是有多少 field，而是具体有多少类别需要预测，这里应该是 9，包括 hair_color 的 5 类，gender 的 2 类，age 的 2 类。
> 
> 已修改，目前已经能正常训练了。谢谢！

请问你是咋修改的呀，还有你的数据集标签是咋样标的呢，能给其中一个图片标签看一看吗？
